modules/distributor/distributor.go

```diff
@@ -223,10 +227,18 @@ func New(cfg Config, clientCfg ingester_client.Config, ingestersRing ring.ReadRi
 
 		subservices = append(subservices, d.generatorsPool)
 
-		d.generatorForwarder = newGeneratorForwarder(logger, reg, d.sendToGenerators, o)
+		d.generatorForwarder = newGeneratorForwarder(logger, d.sendToGenerators, o)
 		subservices = append(subservices, d.generatorForwarder)
 	}
 
+	forwardersManager, err := forwarder.NewManager(d.cfg.Forwarders, logger, o)
```

Comment:
ultimately this calls `grpc.DialContext()` which (i believe) attempts an asynchronous tcp connection. if so, then the code is fine as is.

However, if `grpc.DialContext()` blocks and returns an error on an unsuccessful connection then we need to make changes. i.e. If a forwarder endpoint is down then it is important that a distributor will start successfully.

Can you confirm the behavior either way?

Reply:
Done in [28e63a1](https://github.com/grafana/tempo/pull/1775/commits/28e63a15e5861ff2f40c09d7f48aacca158abec7).